apps/indexer/lib/indexer/fetcher/beacon/deposit.ex

```diff
@@ -0,0 +1,411 @@
+    :last_processed_log_index
+  ]
+
+  def start_link([init_opts, server_opts]) do
+    GenServer.start_link(__MODULE__, init_opts, server_opts)
+  end
+
+  @impl GenServer
+  def init(_opts) do
+    Logger.metadata(fetcher: :beacon_deposit)
+
+    {:ok, nil, {:continue, nil}}
+  end
+
+  @impl GenServer
+  def handle_continue(nil, _state) do
+    chain_id = Application.get_env(:indexer, :chain_id)
+
+    case Client.get_spec() do
+      {:ok,
+       %{
+         "data" => %{
+           "DEPOSIT_CHAIN_ID" => ^chain_id,
+           "DEPOSIT_CONTRACT_ADDRESS" => deposit_contract_address_hash,
+           "DOMAIN_DEPOSIT" => "0x" <> domain_deposit_hex,
+           "GENESIS_FORK_VERSION" => "0x" <> genesis_fork_version_hex
+         }
+       }} ->
+        last_processed_deposit = Deposit.get_latest_deposit() || %{index: -1, block_number: -1, log_index: -1}
+
+        state = %__MODULE__{
+          interval: Application.get_env(:indexer, __MODULE__)[:interval],
+          batch_size: Application.get_env(:indexer, __MODULE__)[:batch_size],
+          deposit_contract_address_hash: deposit_contract_address_hash,
+          domain_deposit: Base.decode16!(domain_deposit_hex, case: :mixed),
+          genesis_fork_version: Base.decode16!(genesis_fork_version_hex, case: :mixed),
+          deposit_index: last_processed_deposit.index,
+          last_processed_log_block_number: last_processed_deposit.block_number,
+          last_processed_log_index: last_processed_deposit.log_index
+        }
+
+        Process.send_after(self(), :process_logs, state.interval)
+
+        {:noreply, state}
+
+      {:ok,
+       %{
+         "data" => %{
+           "DEPOSIT_CHAIN_ID" => chain_id,
+           "DEPOSIT_CONTRACT_ADDRESS" => _deposit_contract_address_hash,
+           "DOMAIN_DEPOSIT" => "0x" <> _domain_deposit_hex,
+           "GENESIS_FORK_VERSION" => "0x" <> _genesis_fork_version_hex
+         }
+       }} ->
+        Logger.error("Misconfigured CHAIN_ID or INDEXER_BEACON_RPC_URL, CHAIN_ID from the node: #{inspect(chain_id)}")
+        {:stop, :wrong_chain_id, nil}
+
+      {:ok, data} ->
+        Logger.error("Unexpected format on beacon spec endpoint: #{inspect(data)}")
+        {:stop, :unexpected_format, nil}
+
+      {:error, reason} ->
+        Logger.error("Failed to fetch beacon spec: #{inspect(reason)}")
+        {:stop, :fetch_failed, nil}
+    end
+  end
+
+  @impl GenServer
+  def handle_cast({:lost_consensus, block_number}, %__MODULE__{} = state) do
+    {_deleted_deposits_count, deleted_deposits} =
+      Repo.delete_all(
+        from(d in Deposit, where: d.block_number >= ^block_number, select: d.index),
+        timeout: :infinity
+      )
+
+    deposit_index = Enum.min(deleted_deposits, fn -> state.deposit_index + 1 end)
+
+    {:noreply,
+     %{
+       state
+       | deposit_index: deposit_index - 1,
+         last_processed_log_block_number: block_number - 1,
+         last_processed_log_index: -1
```

Comment:
I think it should be `(block_number, -1)` or `(block_number - 1, 999..999)`, for paginated fetched to work correctly?

```suggestion
         last_processed_log_block_number: block_number,
         last_processed_log_index: -1
```